Library/Homebrew/cmd/--env.rb

```diff
@@ -1,20 +1,128 @@
 #:  * `--env`:
 #:    Show a summary of the Homebrew build environment.
 
+
 require "extend/ENV"
 require "build_environment"
 
+
 module Homebrew
+  
+  MAX_ITERATIONS = 10
+  VALID_SHELLS = %w(
+    sh csh tcsh bash ksh mksh oksh
+    pdksh zsh fish
+  )
+  DEFAULT_SHELL = "sh"
+
+
+  def canonicalize_shell(path)
+    # basename and strip leading '-' in login shell
+    name = File.basename path
+
+    if VALID_SHELLS.include? name
+      return name
+    end
+    if name.start_with? "-"
+      stripped_name = name.sub(/^[-]/, "")
+      if VALID_SHELLS.include? stripped_name
+        return stripped_name
+      end
+    end
+    return nil
+  end
+
+
+  def get_explicit_shell
+    # parse shell from command line arguments
+    split_arg_index = ARGV.index "--shell"
+    if split_arg_index
+      if split_arg_index + 1 < ARGV.length
+        return ARGV[split_arg_index + 1]
+      else
+        STDERR.puts "--shell option requires argument"
+        exit
+      end
+    end
+
+    long_arg_index = ARGV.index {|s| s.start_with? "--shell="}
+    if long_arg_index
+      shell_name = ARGV[long_arg_index].sub(/^[-][-]shell=/, "")
+      if shell_name == ""
+        return ""
+      end
+
+      canonical = canonicalize_shell(shell_name)
+      if canonical
+        return canonical
+      else
+        STDERR.puts "argument #{shell_name} to --shell option is not a valid shell"
+        exit
+      end
+    end
+    
+    return nil
+  end
+
+
+  def get_ancestor_shell
+    # get the command name for the
+    # nearest parent, grandparent, etc. process
+    # that looks like a shell
+    candidate_pid = Process.ppid.to_s
+
+    for _ in 1..MAX_ITERATIONS
```

Comment:
I disagree. (It was me who asked OP to do it this way instead of using `$SHELL` in the first place.) We know there are some users who switch between shells at run time, including me. And especially if they run in to problems: we're likely to ask them to try things out in `bash` instead of their main shell.

And just from experience working with Oh My Zsh, `$SHELL` is often "wrong" about what shell is currently running. It's like `$EDITOR`: tells you what to run, but not what is running. If we're going to detect at all, we should do it this way.

If including this much code is a problem, then let's fall back to defaulting to bash/Bourne style, and requiring an explicit `--shell` for other shells, but not use `$SHELL`.